### PR TITLE
Use tempconfig for every scene render

### LINKED ISSUE #2307: Renderer only uses one rendering mode for all Scenes in a file
## Description of bug / unexpected behavior

When running `manim animations.py -ql -a`, all of the Scenes are rendered as _either_ .png files _or_ .mp4 files.
For example, if the first Scene in 'animations.py' has no animation, then manim will decide to render that Scene to a .png.
However, then if the next Scene has some animation, then manim will not intelligently switch to rendering .mp4, and instead will produce a .png containing the last frame of the intended animation.

## Expected behavior

If there are some Scenes with animations and some still Scenes in a file, when rendering all Scenes at once, manim should decide for each Scene whether to render to a .png or to an .mp4, based on whether there is animation or not.

## How to reproduce the issue

<details><summary>Code for reproducing the problem</summary>

```py
Paste your code here.
```

</details>

## Additional media files

<details><summary>Images/GIFs</summary>

</details>

## Logs
<details><summary>Terminal output</summary>

```
PASTE HERE OR PROVIDE LINK TO https://pastebin.com/ OR SIMILAR
```

</details>

## System specifications

<details><summary>System Details</summary>

- OS (with version, e.g Windows 10 v2004 or macOS 10.15 (Catalina)):
- RAM:
- Python version (`python/py/python3 --version`):
- Installed modules (provide output from `pip list`):
```
PASTE HERE
```
</details>

<details><summary>LaTeX details</summary>

+ LaTeX distribution (e.g. TeX Live 2020):
+ Installed LaTeX packages:

</details>

<details><summary>FFMPEG</summary>

Output of `ffmpeg -version`:

```
PASTE HERE
```
</details>

### LINKED ISSUE #2470: CLI doesn't use correct formats when using `-a`
## Description of bug / unexpected behavior
When using `manim -a` on a file that contains videos, but also images, all scenes that are rendered after the image is rendered are also rendered as an image.

## Expected behavior
Manim should render the correct format for each scene separately.

## How to reproduce the issue
When executing `manim -a test.py`, it should create `media/videos/test/1080p60/Scene1Video.mp4`, `media/videos/test/1080p/Scene3Video.mp4`, `media/images/test/Scene2Image.png`.

Instead, it creates `media/videos/test/1080p60/Scene1Video.mp4`, `media/images/test/Scene3Video.png`, `media/images/test/Scene2Image.png`.

<details><summary>Code for reproducing the problem</summary>

`test.py`:
```py
from manim import Scene, Circle, Create


class Scene1Video(Scene):
    def construct(self):
        self.play(Create(Circle()))


class Scene2Image(Scene):
    def construct(self):
        self.add(Circle())


class Scene3Video(Scene):
    def construct(self):
        self.play(Create(Circle()))
```

</details>

## System specifications

<details><summary>System Details</summary>

- OS (with version, e.g Windows 10 v2004 or macOS 10.15 (Catalina)): Windows 10 21H2
- RAM: 32GB
- Python version (`python/py/python3 --version`): 3.8.10
- Installed modules (provide output from `pip list`):
```
Package             Version
------------------- -----------
certifi             2021.10.8
charset-normalizer  2.0.7
click               8.0.3
click-default-group 1.2.2
cloup               0.7.1
colorama            0.4.4
colour              0.1.5
commonmark          0.9.1
decorator           5.1.0
glcontext           2.3.4
idna                3.3
isosurfaces         0.1.0
manim               0.14.0
ManimPango          0.4.0.post0
mapbox-earcut       0.12.10
moderngl            5.6.4
moderngl-window     2.4.0
multipledispatch    0.6.0
networkx            2.6.3
numpy               1.21.2
Pillow              8.3.2
pip                 21.3.1
pycairo             1.20.1
pydub               0.25.1
pyglet              1.5.21
Pygments            2.10.0
pyrr                0.10.3
requests            2.26.0
rich                10.12.0
scipy               1.7.1
screeninfo          0.6.7
setuptools          58.2.0
six                 1.16.0
skia-pathops        0.7.1
srt                 3.5.0
tqdm                4.62.3
urllib3             1.26.7
watchdog            2.1.6
```
</details>

### PR DESCRIPTION
**Pull request by Kiran-Raj-Dev**
## Overview: What does this pull request change?

Fixes #2470 and Fixes #2307

## Motivation and Explanation: Why and how do your changes improve the library?

Rendering the last frame changes the `config.save_last_frame` to True globally. And this gets propagatted when rendering other classes within the same script. Using a tempconfig for each scene render solves the problem.

~~Changes `cloup.command` to `click.command`:
both are same but cloup has support for `@constraint` which we don't use anyway. It also stops pyright from complaining that `command` is not exported from cloup.~~

See janluke's review comment below.

Extends the -a flag test to accomodate classes which contain both images and videos

## Further Information and Comments
The tests will fail because of how file logging is setup (directly from `_config.utils`).
I have tested this on #2546 and the tests pass there. So, this PR needs to be merged after that one.

## Reviewer Checklist
- [ ] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [ ] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [ ] If applicable: newly added functions and classes are tested

### CODE REVIEW

**Comment by Darylgolden** [comment]
Converted to draft, feel free to unconvert once tests pass.

**Comment by Kiran-Raj-Dev** [comment]
Well, I already mentioned why the tests would fail

**Comment by Darylgolden** [comment]
> Well, I already mentioned why the tests would fail

We could unconvert after that PR is merged then

**Comment by janluke** [comment]
> Changes cloup.command to click.command:
both are same but cloup has support for @constraint which we don't use anyway. It also stops pyright from complaining that command is not exported from cloup.

Cloup provides support to option groups, which is the reason it was used in the first place. click.command won't fail, but you won't have separate help sections for different groups. 

I see Manim is still using Cloup v0.7.1 when latest version is v0.13.0. Consider upgrading and making use of new features.

File: manim/cli/render/commands.py [hunk 1]
```diff
@@ -24,7 +24,7 @@
 from .render_options import render_options
 
 
-@cloup.command(
+@click.command(
```
[the remark this answers is not in the document]
Comment (Kiran-Raj-Dev) [reply manim/cli/render/commands.py:27]: I tried the following commands:

```zsh
manim --help
manim render --help
```

On this branch:

```
Manim Community v0.15.1

Usage: manim [OPTIONS] COMMAND [ARGS]...

  Animation engine for explanatory math videos

Options:
  --version  Show version and exit.
  --help     Show this message and exit.

Commands:
  render*  Render SCENE(S) from the input FILE.
  cfg      Manages Manim configuration files.
  init     Sets up a new project in current working directory with default
           settings.
           
           It copies files from templates directory and pastes them in the
           current working dir.
  new      Create a new project or insert a new scene.
  plugins  Manages Manim plugins.

  Made with <3 by Manim Community developers.

------------------------------------------------------------------------------------

Manim Community v0.15.1

Usage: manim render [OPTIONS] FILE [SCENE_NAMES]...

  Render SCENE(S) from the input FILE.

  FILE is the file path of the script or a config file.

  SCENES is an optional list of scenes in the file.

Options:
  -c, --config_file TEXT          Specify the configuration file to use for
                                  render settings.
  --custom_folders                Use the folders defined in the
                                  [custom_folders] section of the config file
                                  to define the output folder structure.
  --disable_caching               Disable the use of the cache (still
                                  generates cache files).
  --flush_cache                   Remove cached partial movie files.
  --tex_template TEXT             Specify a custom TeX template file.
  -v, --verbosity [DEBUG|INFO|WARNING|ERROR|CRITICAL]
                                  Verbosity of CLI output. Changes ffmpeg log
                                  level unless 5+.
  --notify_outdated_version / --silent
                                  Display warnings for outdated installation.
  --enable_gui                    Enable GUI interaction.
  --gui_location TEXT             Starting location for the GUI.
  --fullscreen                    Expand the window to its maximum possible
                                  size.
  --enable_wireframe              Enable wireframe debugging mode in opengl.
  --force_window                  Force window to open when using the opengl
                                  renderer, intended for debugging as it may
                                  impact performance
  --dry_run                       Renders animations without outputting image
                                  or video files and disables the window
  -o, --output_file TEXT          Specify the filename(s) of the rendered
                                  scene(s).
  -0, --zero_pad INTEGER RANGE    Zero padding for PNG file names.  [0<=x<=9]
  --write_to_movie                Write the video rendered with opengl to a
                                  file.
  --media_dir PATH                Path to store rendered videos and latex.
  --log_dir PATH                  Path to store render logs.
  --log_to_file                   Log terminal output to file.
  -n, --from_animation_number TEXT
                                  Start rendering from n_0 until n_1. If n_1
                                  is left unspecified, renders all scenes
                                  after n_0.
  -a, --write_all                 Render all scenes in the input file.
  --format [png|gif|mp4|webm|mov]
  -s, --save_last_frame
  -q, --quality [l|m|h|p|k]       Render quality at the follow resolution
                                  framerates, respectively: 854x480 15FPS,
                                  1280x720 30FPS, 1920x1080 60FPS, 2560x1440
                                  60FPS, 3840x2160 60FPS
  -r, --resolution TEXT           Resolution in (W,H) for when 16:9 aspect
                                  ratio isn't possible.
  --fps, --frame_rate FLOAT       Render at this frame rate.
  --renderer [cairo|opengl]       Select a renderer for your Scene.
  --use_opengl_renderer           Render scenes using OpenGL (Deprecated).
  -g, --save_pngs                 Save each frame as png (Deprecated).
  -i, --save_as_gif               Save as a gif (Deprecated).
  --save_sections                 Save section videos in addition to movie
                                  file.
  -s, --save_last_frame           Save last frame as png (Deprecated).
  -t, --transparent               Render scenes with alpha channel.
  --use_projection_fill_shaders   Use shaders for OpenGLVMobject fill which
                                  are compatible with transformation matrices.
  --use_projection_stroke_shaders
                                  Use shaders for OpenGLVMobject stroke which
                                  are compatible with transformation matrices.
  --progress_bar [display|leave|none]
                                  Display progress bars and/or keep them
                                  displayed.
  -p, --preview                   Preview the Scene's animation. OpenGL does a
                                  live preview in a popup window. Cairo opens
                                  the rendered video file in the system
                                  default media player.
  -f, --show_in_file_browser      Show the output file in the file browser.
  --jupyter                       Using jupyter notebook magic.
  --help                          Show this message and exit.

  Made with <3 by Manim Community developers.
```

On Main:

```
Manim Community v0.15.1

Usage: manim [OPTIONS] COMMAND [ARGS]...

  Animation engine for explanatory math videos

Options:
  --version  Show version and exit.
  --help     Show this message and exit.

Commands:
  render*  Render SCENE(S) from the input FILE.
  cfg      Manages Manim configuration files.
  init     Sets up a new project in current working directory with default
           settings.
           
           It copies files from templates directory and pastes them in the
           current working dir.
  new      Create a new project or insert a new scene.
  plugins  Manages Manim plugins.

  Made with <3 by Manim Community developers.
  
--------------------------------------------------------------------

Manim Community v0.15.1

Usage: manim render [OPTIONS] FILE [SCENE_NAMES]...

  Render SCENE(S) from the input FILE.

  FILE is the file path of the script or a config file.

  SCENES is an optional list of scenes in the file.

Global options:
  -c, --config_file TEXT          Specify the configuration file to use for
                                  render settings.
  --custom_folders                Use the folders defined in the
                                  [custom_folders] section of the config file
                                  to define the output folder structure.
  --disable_caching               Disable the use of the cache (still
                                  generates cache files).
  --flush_cache                   Remove cached partial movie files.
  --tex_template TEXT             Specify a custom TeX template file.
  -v, --verbosity [DEBUG|INFO|WARNING|ERROR|CRITICAL]
                                  Verbosity of CLI output. Changes ffmpeg log
                                  level unless 5+.
  --notify_outdated_version / --silent
                                  Display warnings for outdated installation.
  --enable_gui                    Enable GUI interaction.
  --gui_location TEXT             Starting location for the GUI.
  --fullscreen                    Expand the window to its maximum possible
                                  size.
  --enable_wireframe              Enable wireframe debugging mode in opengl.
  --force_window                  Force window to open when using the opengl
                                  renderer, intended for debugging as it may
                                  impact performance
  --dry_run                       Renders animations without outputting image
                                  or video files and disables the window

Output options:
  -o, --output_file TEXT          Specify the filename(s) of the rendered
                                  scene(s).
  -0, --zero_pad INTEGER RANGE    Zero padding for PNG file names.  [0<=x<=9]
  --write_to_movie                Write the video rendered with opengl to a
                                  file.
  --media_dir PATH                Path to store rendered videos and latex.
  --log_dir PATH                  Path to store render logs.
  --log_to_file                   Log terminal output to file.

Render Options:
  -n, --from_animation_number TEXT
                                  Start rendering from n_0 until n_1. If n_1
                                  is left unspecified, renders all scenes
                                  after n_0.
  -a, --write_all                 Render all scenes in the input file.
  --format [png|gif|mp4|webm|mov]
  -s, --save_last_frame
  -q, --quality [l|m|h|p|k]       Render quality at the follow resolution
                                  framerates, respectively: 854x480 15FPS,
                                  1280x720 30FPS, 1920x1080 60FPS, 2560x1440
                                  60FPS, 3840x2160 60FPS
  -r, --resolution TEXT           Resolution in (W,H) for when 16:9 aspect
                                  ratio isn't possible.
  --fps, --frame_rate FLOAT       Render at this frame rate.
  --renderer [cairo|opengl]       Select a renderer for your Scene.
  --use_opengl_renderer           Render scenes using OpenGL (Deprecated).
  -g, --save_pngs                 Save each frame as png (Deprecated).
  -i, --save_as_gif               Save as a gif (Deprecated).
  --save_sections                 Save section videos in addition to movie
                                  file.
  -s, --save_last_frame           Save last frame as png (Deprecated).
  -t, --transparent               Render scenes with alpha channel.
  --use_projection_fill_shaders   Use shaders for OpenGLVMobject fill which
                                  are compatible with transformation matrices.
  --use_projection_stroke_shaders
                                  Use shaders for OpenGLVMobject stroke which
                                  are compatible with transformation matrices.

Ease of access options:
  --progress_bar [display|leave|none]
                                  Display progress bars and/or keep them
                                  displayed.
  -p, --preview                   Preview the Scene's animation. OpenGL does a
                                  live preview in a popup window. Cairo opens
                                  the rendered video file in the system
                                  default media player.
  -f, --show_in_file_browser      Show the output file in the file browser.
  --jupyter                       Using jupyter notebook magic.

Other options:
  --help                          Show this message and exit.

  Made with <3 by Manim Community developers.
```

Seems to work fine on both. Am I missing something?

[the remark this answers is not in the document]
Comment (janluke) [reply manim/cli/render/commands.py:27]: The answer is already in my previous two comments ^^. The first help (with click) has a unique list of options titled "Options". The second (cloup) have options divided in multiple sections (Global options, Render options etc), i.e. options groups. As I wrote, the program won't break. You just silently lose the "option groups" feature in the --help text.

[the remark this answers is not in the document]
Comment (Kiran-Raj-Dev) [reply manim/cli/render/commands.py:27]: Ah! I see what you mean. I think its a nice feature to have.
 I'll revert that change 👍

**Comment by Kiran-Raj-Dev** [comment]
> Converted to draft, feel free to unconvert once tests pass.

@Darylgolden All tests pass now ^^

File: tests/test_scene_rendering/infallible_scenes.py [hunk 2]
```diff
@@ -1,13 +1,18 @@
...
-from manim import *
+from manim import Scene, Square
 
 
 class Wait1(Scene):
     def construct(self):
         self.wait()
 
 
+class Wait3(Scene):
```
Comment (behackl) [inline tests/test_scene_rendering/infallible_scenes.py:11]: this seems unrelated?

**Review by behackl — CHANGES_REQUESTED** [review]
the idea and implementation seems fine to me, but i feel like the behavior should be documented too (ideally also tested, but setting up a meaningful cli-test for this might not be straightforward; perhaps it would work with a separate test file in which in the construct method of one scene ˋconfig.some_setting = 42ˋ is run, and in se second scene rendered immediately after the construct method has an assert checking that the value has been reset / not set at all).

what do you think?

File: tests/test_scene_rendering/infallible_scenes.py (hunk 2 above)
Comment (Kiran-Raj-Dev) [reply tests/test_scene_rendering/infallible_scenes.py:11]: Have you taken a look at the issues?
The thing was that if such a thing ran on main now then the output you get is not expected, `Wait1` will be a video but `Wait3` and `Wait2` will be images, whereas `Wait2` should actually be a video.

So, this is kind of a test to check that 😓

**Comment by MrDiver** [comment]
Just undraft it when it's ready for review again

File: tests/test_scene_rendering/infallible_scenes.py (hunk 2 above)
Comment (behackl) [reply tests/test_scene_rendering/infallible_scenes.py]: Now I actually understand -- also I appreciate that you changed the order of the test scenes. :-)
